REGRESSION (263722@main): Unable to scroll language picker on etrade.com

whsieh opened the pull request:
#### ac10daa79e26b4f062cb2c1842b4787698574899
<pre>
REGRESSION (263722@main): Unable to scroll language picker on etrade.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=258433">https://bugs.webkit.org/show_bug.cgi?id=258433</a>
rdar://109613133

Reviewed by Simon Fraser.

After the changes in 263722@main, certain overflow scrolling containers on etrade.com are no longer
scrollable via pan gesture or mouse wheel on iOS; this is because we fail to create corresponding
native child scroll views corresponding to these regions in the view hierarchy, which in turn is
because we never end up setting `m_hasCompositedScrollableOverflow` to `true` inside of
`RenderLayerScrollableArea::computeHasCompositedScrollableOverflow()`, despite the fact that we have
vertical scrollable overflow and we can use composited scrolling.

The overflow scrolling container in question is shown by removing `visibility: hidden;` on an
ancestor of these containers, which currently triggers only a style update (calling into the above
method with `LayoutUpToDate::No`) with no subsequent layout update, causing
`m_hasCompositedScrollableOverflow` to remain `false`.

To address this, we make an adjustment in `RenderLayer::styleChanged` to pass `LayoutUpToDate::Yes`
into `computeHasCompositedScrollableOverflow`, in the case where the style difference only triggers
at most a layer repaint (or recomposite). In the case where we&apos;re only issuing a repaint as a result
of the style change (e.g. when toggling visibility), the overflow amount isn&apos;t going to change, so
it&apos;s safe to compute `m_hasCompositedScrollableOverflow` then and there. However, in the case where
the style difference is going to trigger layout, it&apos;s safe to just run the computation using
`LayoutUpToDate::No`, since we&apos;ll eventually recompute it with `LayoutUpToDate::Yes` after we finish
performing layout.

* LayoutTests/fast/scrolling/ios/overflow-scroll-after-visibility-change-expected.txt: Added.
* LayoutTests/fast/scrolling/ios/overflow-scroll-after-visibility-change.html: Added.

Add a layout test to verify that we propagate overflow scrolling regions via scrolling state tree to
the UI process in this scenario, after toggling `visibility` to `visible`.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):

Canonical link: <a href="https://commits.webkit.org/265624@main">https://commits.webkit.org/265624@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/995640f3819c9200419a9b656cb95e4f7cb10ae7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11349 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11559 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11899 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12998 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10816 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11370 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13939 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11543 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13731 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11511 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12385 "2 flakes 3 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9607 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13417 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9680 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10286 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17481 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10749 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10443 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13661 "run-api-tests-without-change (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10861 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8942 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10033 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/10177 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2746 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14307 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10716 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1356,6 +1356,23 @@ bool RenderStyle::changeRequiresRecompositeLayer(const RenderStyle& other, Optio
+
+bool RenderStyle::changeRequiresSimplifiedLayout(const RenderStyle& other, OptionSet<StyleDifferenceContextSensitiveProperty>&) const
+{
+    if (mayRequireOverflowScrolling(*this) && mayRequireOverflowScrolling(other)) {
```
I wish there was a diff value in-between StyleDifference::SimplifiedLayout and StyleDifference::RepaintLayer, something along the lines of "skip layout but run a full layer update" and initiate that unconditionally on visibility change (I am somewhat convinced this helps with resetting some stale cached states in a non-scrolling case too)

whsieh replied:
Oh, interesting — I had thought that "skip layout but run a full layer update" was already basically `SimplifiedLayout`, but I suppose `SimplifiedLayout` still triggers some kind of layout... Are you proposing something like a `RepaintLayerAndRecomputeOverflow`?

whsieh replied:
(Filed https://bugs.webkit.org/show_bug.cgi?id=258522 for this)

smfr commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1356,6 +1356,23 @@ bool RenderStyle::changeRequiresRecompositeLayer(const RenderStyle& other, Optio
     return false;
 }
 
+inline static bool mayRequireOverflowScrolling(const RenderStyle& style)
```
I think it would be preferable to just skip the overflow check, and trigger simplified layout any time visibility changes.

whsieh committed:
REGRESSION (263722@main): Unable to scroll language picker on etrade.com
https://bugs.webkit.org/show_bug.cgi?id=258433
rdar://109613133

Reviewed by Simon Fraser.

After the changes in 263722@main, certain overflow scrolling containers on etrade.com are no longer
scrollable via pan gesture or mouse wheel on iOS; this is because we fail to create corresponding
native child scroll views corresponding to these regions in the view hierarchy, which in turn is
because we never end up setting `m_hasCompositedScrollableOverflow` to `true` inside of
`RenderLayerScrollableArea::computeHasCompositedScrollableOverflow()`, despite the fact that we have
vertical scrollable overflow and we can use composited scrolling.

The overflow scrolling container in question is shown by removing `visibility: hidden;` on an
ancestor of these containers, which currently triggers only a style update (calling into the above
method with `LayoutUpToDate::No`) with no subsequent layout update, causing
`m_hasCompositedScrollableOverflow` to remain `false`.

To address this, we make an adjustment in `RenderLayer::styleChanged` to pass `LayoutUpToDate::Yes`
into `computeHasCompositedScrollableOverflow`, in the case where the style difference only triggers
at most a layer repaint (or recomposite). In the case where we're only issuing a repaint as a result
of the style change (e.g. when toggling visibility), the overflow amount isn't going to change, so
it's safe to compute `m_hasCompositedScrollableOverflow` then and there. However, in the case where
the style difference is going to trigger layout, it's safe to just run the computation using
`LayoutUpToDate::No`, since we'll eventually recompute it with `LayoutUpToDate::Yes` after we finish
performing layout.

* LayoutTests/fast/scrolling/ios/overflow-scroll-after-visibility-change-expected.txt: Added.
* LayoutTests/fast/scrolling/ios/overflow-scroll-after-visibility-change.html: Added.

Add a layout test to verify that we propagate overflow scrolling regions via scrolling state tree to
the UI process in this scenario, after toggling `visibility` to `visible`.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):

Canonical link: https://commits.webkit.org/265624@main
